config/config-builder/src/full_node_config.rs

```diff
@@ -79,61 +82,78 @@ impl FullNodeConfig {
+        .clone();
+
+        // TODO: Do we need this?
+        public_network.listen_address = self.listen_address.clone();
+
+        Ok(full_node_config)
     }
 
+    // This is to add a VFN network to a validator.
     pub fn extend_validator(&self, config: &mut NodeConfig) -> Result<()> {
-        let (mut configs, _) = self.build_internal(false)?;
-
-        let mut new_net = configs.swap_remove(configs.len() - 1);
-        let seed_config = &new_net
-            .full_node_networks
-            .last()
-            .ok_or(Error::MissingFullNodeNetwork)?;
-        let seed_addrs = generator::build_seed_addrs(&seed_config, self.bootstrap.clone());
-
-        let mut network = new_net.full_node_networks.swap_remove(0);
-        network.discovery_method = DiscoveryMethod::gossip(self.advertised_address.clone());
-        network.listen_address = self.listen_address.clone();
-        network.seed_addrs = seed_addrs;
-        config.full_node_networks.push(network);
+        let full_node_config = self.build()?;
+        let vfn_network = (*networks_with_id(&full_node_config, NetworkId::vfn_network())
+            .first()
+            .unwrap())
+        .clone();
+        for net in &config.full_node_networks {
+            let new_net_id = vfn_network.peer_id();
+            let net_id = net.peer_id();
+            ensure!(new_net_id != net_id, "Network already exists");
+        }
+        config.full_node_networks.push(vfn_network);
         Ok(())
     }
 
+    // Add a public network to a full node
+    // TODO: This doesn't make sense anymore since the network Id would be repeated, the interface should take more info
     pub fn extend(&self, config: &mut NodeConfig) -> Result<()> {
-        let mut full_node_config = self.build()?;
-        let new_net = full_node_config.full_node_networks.swap_remove(0);
+        let full_node_config = self.build()?;
+        let mut new_public_network = (*networks_with_id(&full_node_config, NetworkId::Public)
+            .first()
+            .unwrap())
+        .clone();
+
+        // TODO: this seems like it shouldn't matter to be exactly the same
+        new_public_network.listen_address = self.listen_address.clone();
         for net in &config.full_node_networks {
-            let new_net_id = new_net.peer_id();
+            let new_net_id = new_public_network.peer_id();
             let net_id = net.peer_id();
             ensure!(new_net_id != net_id, "Network already exists");
         }
-        config.full_node_networks.push(new_net);
+        config.full_node_networks.push(new_public_network);
         Ok(())
     }
 
-    fn build_internal(
-        &self,
-        randomize_ports: bool,
-    ) -> Result<(Vec<NodeConfig>, Ed25519PrivateKey)> {
+    /// Builds a swarm of nodes based on the number of full nodes
+    /// The last node will always be the ValidatorFullNode, and the rest are public
+    fn build_swarm_internal(&self, randomize_ports: bool) -> Result<Vec<NodeConfig>> {
```

Comment:
i think we can actually delete this whole method, since build_swarm is now unused. hopefully this can simplify some things : p

i would also delete the build_swarm in config-builder validator_config